codec/src/main/java/io/netty/handler/codec/BatchLengthFieldBasedFrameDecoder.java

```diff
@@ -0,0 +1,136 @@
+     * {@link TooLongFrameException} will be thrown.
+     * @param lengthFieldOffset the offset of the length field
+     * @param lengthFieldLength the length of the length field
+     */
+    public BatchLengthFieldBasedFrameDecoder(int startBufferSize, int maxFrameLength, int lengthFieldOffset,
+                                             int lengthFieldLength) {
+        super(startBufferSize);
+
+        if (maxFrameLength <= 0) {
+            throw new IllegalArgumentException(
+                    "maxFrameLength must be a positive integer: " +
+                    maxFrameLength);
+        }
+
+        if (lengthFieldOffset < 0) {
+            throw new IllegalArgumentException(
+                    "lengthFieldOffset must be a non-negative integer: " +
+                    lengthFieldOffset);
+        }
+
+        if (lengthFieldOffset > maxFrameLength - lengthFieldLength) {
+            throw new IllegalArgumentException(
+                    "maxFrameLength (" + maxFrameLength + ") " +
+                    "must be equal to or greater than " +
+                    "lengthFieldOffset (" + lengthFieldOffset + ") + " +
+                    "lengthFieldLength (" + lengthFieldLength + ").");
+        }
+
+        this.maxFrameLength = maxFrameLength;
+        this.lengthFieldOffset = lengthFieldOffset;
+        this.lengthFieldLength = lengthFieldLength;
+        lengthFieldEndOffset = lengthFieldOffset + lengthFieldLength;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) {
+        int readableBytes = buffer.readableBytes();
+        int rdIdx = buffer.readerIndex();
+
+        while (readableBytes >= lengthFieldEndOffset) {
+            int actualLengthFieldOffset = rdIdx + lengthFieldOffset;
+            long frameLength =
+                    UnadjustedFrameLengthDecoderUtil.decode(buffer, actualLengthFieldOffset, lengthFieldLength);
+
+            if (frameLength < 0) {
+                throw new CorruptedFrameException(
+                        "negative pre-adjustment length field: " + frameLength);
+            }
+
+            frameLength += lengthFieldEndOffset;
+
+            if (frameLength > maxFrameLength) {
+                throw new TooLongFrameException(
+                        "Adjusted frame length exceeds " + maxFrameLength +
+                        ": " + frameLength + " - discarded");
+            }
+
+            int frameLengthInt = (int) frameLength;
+
+            if (readableBytes < frameLengthInt) {
+                break;
+            }
+
+            readableBytes -= frameLengthInt;
+            rdIdx += frameLengthInt;
+        }
+
+        if (rdIdx == 0) {
+            return;
+        }
+
+        final int previousWriterIndex = buffer.writerIndex();
+
+        buffer.writerIndex(rdIdx);
+
+        final ByteBuf chunk = ctx.alloc().buffer(rdIdx, rdIdx).writeBytes(buffer, rdIdx);
```

Comment:
Why it may not be 0 based? [BatchFrameDecoder](https://github.com/netty/netty/pull/6052/files#diff-db29166cae2ff3be4fe271a0b1311ed1R88) sets the reader index to 0 after every `decode` method invocation.